iroh-rpc/src/client.rs

```diff
@@ -0,0 +1,313 @@
+        };
+        // examine header here and determine if we want to accept file
+        Ok(InStream::new(header, stream, self.command_sender.clone()))
+    }
+
+    /// Get the peer id from the Client's address book, based on the namespace
+    fn get_peer_id(&self, namespace: &str) -> Result<PeerId, RpcError> {
+        match self.addresses.get(namespace) {
+            Some((_, id)) => Ok(*id),
+            None => Err(RpcError::NamespaceNotFound(namespace.into())),
+        }
+    }
+
+    /// Get the multiaddr from the Client's address book, based on the namespace
+    fn get_multiaddr(&self, namespace: &str) -> Result<Multiaddr, RpcError> {
+        match self.addresses.get(namespace) {
+            Some((addr, _)) => Ok(addr.clone()),
+            None => Err(RpcError::NamespaceNotFound(namespace.into())),
+        }
+    }
+}
+
+struct Iter {
+    num: u64,
+}
+
+impl Iter {
+    fn new() -> Self {
+        Iter { num: 0 }
+    }
+}
+
+impl Iterator for Iter {
+    type Item = u64;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.num += 1;
+        Some(self.num)
+    }
+}
+
+/// Dial a single address
+async fn dial(
+    mut command_sender: mpsc::Sender<Command>,
+    addr: Multiaddr,
+    peer_id: PeerId,
+) -> Result<(), RpcError> {
+    let (sender, receiver) = oneshot::channel();
+    command_sender
+        .send(Command::Dial {
+            peer_id,
+            peer_addr: addr,
+            sender,
+        })
+        .await
+        .expect("Receiver to not be dropped.");
+    match receiver.await.expect("Sender to not be dropped.") {
+        SenderType::Ack => Ok(()),
+        SenderType::Error(_) => Err(RpcError::TODO),
+        _ => Err(RpcError::TODO),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::sync::Arc;
+
+    use libp2p::identity::Keypair;
+
+    use crate::config::RpcConfig;
+    use crate::handler;
+    use crate::serde::{deserialize_request, serialize_response, Deserialize};
+    use crate::swarm;
+
+    #[derive(Serialize, Debug, Clone)]
+    struct GetParams {
+        resource_id: String,
+    }
+
+    #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+    struct GetPayload {
+        response: String,
+    }
+
+    struct Get;
+
+    // DIG! I am struggling
+    // I'm also not sure if I just don't know what's going on, or if this is a bad arrangement (or
+    // both)
+    // Trying to implement Factory for Get, implying that it will be the `Get` handler that is added
+    // to one of the test rpc clients. Namespace "a", method "get", to be added to the b_rpc_config
+    // using
+    // `b_rpc_config.with_namespace("a", |namespace| {
+    //     namespace.with_method("get", Get)
+    // })`
+    // Confused about some things that aren't compiling & also how we ensure that `handler::State<T>`
+    // is a String in this case, is there a way to connect the State concrete type with the
+    // implementation of Get?
+    // anyway, would love some imput!
```

Comment:
Help plz @dignifiedquire

Reply:
There was a little bit more to do, let me know if you want to chat about it

https://github.com/n0-computer/iroh/commit/8e0fabb1ebe3fe9feba51a70dc424b29af6776b1